File: 844_backspace_string_compare.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var backspaceCompare = function(s, t) {
    const storageS = [];
    const storageT = [];

    stringAnalyzer(s,storageS);

    stringAnalyzer(t,storageT);

    let newS = storageS.join('');
    let newT = storageT.join('');

    if (newS !== newT) return false;
    else return true;
};

var stringAnalyzer = function(string, storage) {
    for (let i = 0; i < string.length; i++){
        if (string[i] === "#"){
            storage.pop();
        } else {
            storage.push(string[i])
        }
    }
}

//Time Complexity O(n), Space Complexity O(n);

File: 844_backspace_string_compare.py
class Solution:
    def stringAnalyzer(self, string: str, storage: list) -> list:
        for i in range(len(string)):
            if string[i] == "#" and len(storage) > 0:
                storage.pop()
            elif string[i] != "#":
                storage.append(string[i])
        return storage

    def backspaceCompare(self, s: str, t: str) -> bool:
        storageS = self.stringAnalyzer(s, [])
        storageT = self.stringAnalyzer(t, [])

        newS = ''.join(storageS)
        newT = ''.join(storageT)

        print(newS)
        print(newT)

        if newS != newT:
            return False
        else:
            return True
